sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +380,483 @@ def transform(self, X):
+        self.min_value = min_value
+        self.max_value = max_value
+        self.verbose = verbose
+        self.random_state = random_state
+
+    def _impute_one_feature(self,
+                            X_filled,
+                            mask_missing_values,
+                            feat_idx,
+                            neighbor_feat_inds,
+                            estimator=None,
+                            min_std=1e-6):
+        """Imputes a single feature from the others provided.
+
+        This function predicts the missing values of one of the features using
+        the current estimates of all the other features. The `estimator` must
+        support `return_std=True` in its ``predict`` method for this function
+        to work.
+
+        Parameters
+        ----------
+        X_filled : array-like
+            Input data with the most recent imputations.
+
+        mask_missing_values : array-like
+            Input data's missing indicator matrix.
+
+        feat_idx : integer
+            Index of the feature currently being imputed.
+
+        neighbor_feat_inds : array-like
+            Indices of the features to be used in imputing `feat_idx`
+
+        estimator : object, default=BayesianRidgeRegression
+            The estimator to use at this step of the round-robin imputation.
+            As of now, it must support `return_std` in its ``predict`` method.
+            If None, it will be fit, otherwise not.
+
+        min_std : float, optional (default=1e-6)
+            The smallest allowable standard deviation for the posterior
+            sampling step.
+
+        Returns
+        -------
+        X_filled : array-like
+            Input data with `X_filled[missing_row_mask, feat_idx]`
+            updated.
+
+        estimator : estimator with sklearn API
+            The fitted estimator used to impute
+            `X_filled[missing_row_mask, feat_idx]`.
+        """
+
+        # if nothing is missing, just return the default
+        rng = check_random_state(self.random_state)
+        if mask_missing_values[:, feat_idx].sum() == 0:
+            return X_filled, estimator
+        missing_row_mask = mask_missing_values[:, feat_idx]
+
+        # if no estimator provided, instantiate a new one and fit
+        if estimator is None:
+            X_train = X_filled[:, neighbor_feat_inds][~missing_row_mask]
+            y_train = X_filled[:, feat_idx][~missing_row_mask]
+            if np.std(y_train) > 0:
+                estimator = clone(self.estimator_)
+                estimator.fit(X_train, y_train)
+            else:
+                estimator = DummyRegressor()
+                estimator.fit(X_train, y_train)
+
+        # get posterior samples
+        X_test = X_filled[:, neighbor_feat_inds][missing_row_mask]
+        mus, sigmas = estimator.predict(X_test, return_std=True)
+        sigmas = np.maximum(sigmas, min_std)
+        imputed_values = rng.normal(mus, sigmas)
```

Comment:
@glemaitre I'm getting test failure here with `ValueError: scale < 0`, which makes no sense since `sigmas` are guaranteed to be at least `min_std`, which is `1e-6`. Any idea what this could mean? It's only failing for SOME of the tests: https://travis-ci.org/scikit-learn/scikit-learn/builds/295487534?utm_source=github_status&utm_medium=notification

Reply:
this is strange. I would think this is related to numpy version. I would check the values of `sigmas` to be sure that there is not something fishy.

Reply:
I am trying to reproduce this on my machine and I can't! Probably should assert that `sigmas` are finite or something?

Reply:
That's weird because `np.inf` or `np.nan` should not trigger this error. I would really print the value and check the values in the CI. I could not reproduce as well (even with the same library version --- but this is late :))